### PR TITLE
InstallScripts: pin teleport version using ServerVersion

### LINKED ISSUE #28113: Race between Cloud upgrade and stable/cloud repos
Right now we have the following problem with Cloud upgrade and packages published to stable/cloud repositories:

- Cloud upgrades to new version, say, 13.1.1.
- Our install scripts (Discover/auto-discovery) start using version 13.1.1 for stable/cloud repositories.
- stable/cloud repos don't have 13.1.1 because it is only published there after all tenants have been upgraded to 13.1.1.

Proposed fix:

- For tenants enrolled in auto-upgrades, keep using `stable/cloud` channel and but use version indicated by S3 endpoint (https://updates.releases.teleport.dev/v1/stable/cloud/version) instead of their cluster version.
- For all other (legacy) tenants, use `stable/vXXX` channels based on their cluster version as it used to be.

### PR DESCRIPTION
**Pull request by marcoandredinis**
When Automatic Upgrades are enabled and it is cloud, it will install teleport using:
- stable/cloud repo channel (yum, apt)
- pin the version to the one present at: https://updates.releases.teleport.dev/v1/stable/cloud/version

Demo:
<details>
  <summary>Running in Cloud with Auto Upgrades</summary>

image

I got the expected channel and version:
image
</details>

<details>
  <summary>Running in Cloud without Auto Upgrades</summary>
image
</details>

<details>
  <summary>An enterprise version running self-hosted with auto upgrades</summary>

image

</details>

Fix https://github.com/gravitational/teleport/issues/28113

### CODE REVIEW

File: lib/automaticupgrades/version.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+const (
+	// stableCloudVersionBaseURL is the base URL for the server that returns the current stable/cloud version.
+	stableCloudVersionBaseURL = "https://updates.releases.teleport.dev"
+
+	// stableCloudVersionURL is the URL that returns the current stable/cloud version.
+	stableCloudVersionURL = "/v1/stable/cloud/version"
```
Comment (r0mant) [inline lib/automaticupgrades/version.go:36]:
```suggestion
	// stableCloudVersionPath is the URL path that returns the current stable/cloud version.
	stableCloudVersionPath = "/v1/stable/cloud/version"
```

File: lib/web/join_tokens.go [hunk 2]
```diff
@@ -392,9 +392,17 @@ func getJoinScript(ctx context.Context, settings scriptSettings, m nodeAPIGetter
...
+	// For Teleport Cloud installations with AutomaticUpgrades enabled, use the `stable/cloud` repo channel and get version from
+	// https://updates.releases.teleport.dev/v1/stable/cloud/version
+	if settings.installUpdater {
 		repoChannel = stableCloudChannelRepo
+		cloudStableVersion, err := automaticupgrades.Version(ctx, settings.automaticUpgradesVersionBaseURL)
+		if err != nil {
+			return "", trace.Wrap(err)
+		}
+
+		// cloudStabelVersion has vX.Y.Z format, however the script expects the version to not include the `v`
```
Comment (r0mant) [inline lib/web/join_tokens.go:404]:
```suggestion
		// cloudStableVersion has vX.Y.Z format, however the script expects the version to not include the `v`
```

File: lib/utils/utils.go [hunk 3]
```diff
@@ -506,11 +505,6 @@ func ReadOrMakeHostUUID(dataDir string) (string, error) {
 	return id, nil
 }
 
-// PrintVersion prints human readable version
-func PrintVersion() {
```
Comment (r0mant) [inline lib/utils/utils.go:510]: This change seems irrelevant to the issue.

File: lib/web/join_tokens.go [hunk 4]
```diff
@@ -392,9 +392,17 @@ func getJoinScript(ctx context.Context, settings scriptSettings, m nodeAPIGetter
 	// By default, it will use `stable/v<majorVersion>`, eg stable/v12
 	repoChannel := ""
 
-	// For Teleport Cloud installations, use the `stable/cloud` channel.
-	if settings.stableCloudChannelRepo {
+	// For Teleport Cloud installations with AutomaticUpgrades enabled, use the `stable/cloud` repo channel and get version from
```
Comment (r0mant) [inline lib/web/join_tokens.go:395]: I would update this comment to explicitly mention that currently auto-upgrades in the install script are enabled _only_ for Cloud customers enrolled in auto-upgrades. Otherwise, it looks strange that we're checking generic "installUpdater" variable but override the channel to the cloud one.

**Review by r0mant — APPROVED** [review]
@marcoandredinis Can you please also test it on a Cloud tenant to make sure we don't run into the same situation during upgrade as we did earlier this week?

File: lib/utils/utils.go (hunk 3 above)
Comment (marcoandredinis) [reply lib/utils/utils.go:510]: This was causing an import cycle because:
```
github.com/gravitational/teleport/lib/modules imports github.com/gravitational/teleport/lib/automaticupgrades imports github.com/gravitational/teleport/lib/utils imports github.com/gravitational/teleport/lib/modules: import cycle not allowed .
```

It was only used in our binaries and it was a one liner, so I just replaced the function call

File: lib/web/join_tokens.go (hunk 4 above)
Comment (marcoandredinis) [reply lib/web/join_tokens.go]: Updated to:
```

	// The install script will install the updater (teleport-ent-updater) for Cloud customers enrolled in Automatic Upgrades.
	// The repo channel used must be `stable/cloud` which has the available packages for the Cloud Customer's agents.
	// It pins the teleport version to the one specified by https://updates.releases.teleport.dev/v1/stable/cloud/version
	// This ensures the initial installed version is the same as the `teleport-ent-updater` would install.
```

**Comment by marcoandredinis** [comment]
> @marcoandredinis Can you please also test it on a Cloud tenant to make sure we don't run into the same situation during upgrade as we did earlier this week?

Yes :+1: 
I tested using an enterprise build with the license from my tenant in Cloud Stage 
But using the real env seems a better approach :+1:

**Comment by marcoandredinis** [comment]
With auto upgrades on 
image

And Server reporting version 14-dev
image

I got the expected channel and version:
image

**Comment by marcoandredinis** [comment]
After disabling automatic upgrades for my tenant, I get the following script
image

Working as expected 👍

**Review by hugoShaka — APPROVED** [review]

**Review by tigrato — APPROVED** [review]

**Review by AntonAM — APPROVED** [review]
